### ISSUE TITLE
Exception in OneTimeSetUp has no stack trace

### ISSUE THREAD

**Issue by provegard** [nunit/nunit#3409]
When an exception is thrown in code called from a `OneTimeSetUp` method, I get:

> Exception doesn't have a stacktrace

Code to demonstrate the problem: 

    public class Tests
    {
        [OneTimeSetUp]
        public void Setup()
        {
            throw new Exception("oops");
        }

        [Test]
        public void Test1()
        {
            Assert.Pass();
        }
    }

Versions:

* .NET Core 3.0
* NUnit 3.12.0
* NUnit3TestAdapter 3.15.1
* Microsoft.NET.Test.SDK 16.3.0

**Comment by CharliePoole** [comment]
Please clarify "I get..." That is... How are you running the tests and where do you see this message?

**Comment by provegard** [comment]
I run them in one of two ways:

1. Visual Studio 2019 with ReSharper. The "Exception doesn't have a stacktrace" message is printed as the test failure ouput:

![image](https://user-images.githubusercontent.com/795010/67704665-d5a0e900-f9b5-11e9-84b1-ccebeef75f2f.png)

2. Using `dotnet vstest` with TRX reporting. In this case the report contains:

```
<UnitTestResult ...>
  <Output>
    <ErrorInfo>
      <Message>OneTimeSetUp: System.Exception : oops</Message>
    </ErrorInfo>
  </Output>
</UnitTestResult>
```

Changing to `SetUp` results in the full stack trace being printed, so it appears to be a problem with `OneTimeSetUp`.

**Comment by CharliePoole** [comment]
If an exception is thrown in OneTimeSetUp, then the tests are not run. Consequently no exception is ever thrown for the individual test case. Instead, it's thrown for the suite or fixture.

Because TestExplorer (and I believe Resharper as well) never reports errors on the fixture, we report the message __as if__ the exception was thrown for the test case itself - otherwise, you would never see it. We do not report the stack trace simply because that would be a lot of repetition and you can generally find where the exception was thrown pretty easily, if not by inspection, then in the debugger.

If you run tests using nunit3-console, we actually report errors at the fixture level as well, so you can see the stack trace there. The same is true if you run using the TestCentric GUI.

IOW, at least in my view, this is a limitation of using Test Explorer, which only treats test methods as tests, not test suites.

**Comment by provegard** [comment]
> We do not report the stack trace simply because that would be a lot of repetition and you can generally find where the exception was thrown pretty easily, if not by inspection, then in the debugger.

Normally, yes. In my case, I was modifying an application to run on both Windows and Linux, and on Linux I got exceptions from GDI-related code ("Parameter is invalid" - not easy to figure that out :) ). I did not have a full IDE available there, only the `dotnet` CLI tooling. As I mentioned, I could change to `SetUp` and see the stack trace, but it would have been nicer to be able to pinpoint the root cause quicker. 

Occasionally, I have intermittent test failures on the build server, and in those cases the suppressed stack trace makes it very hard to know what's going on.

> IOW, at least in my view, this is a limitation of using Test Explorer, which only treats test methods as tests, not test suites.

Forgive me if I misunderstand, but if I interpret you correctly it's NUnit that suppresses the stack trace, and since I observe that an exception has occurred both in Test Explorer/ReSharper and in a test report, it seems to me that the problem lies with NUnit. Or rather, that the only place where a fix is possible is within NUnit.

**Comment by provegard** [comment]
Btw, I agree that repeating the stack trace for each test is a lot of repetetion. However, if I use `SetUp` that is exactly what happens - I see the exception will full stack trace for each test. With `OneTimeSetUp`, I see the exception without stack trace for each test. I'd argue that stack trace repetition is better than no stack trace.

**Comment by CharliePoole** [comment]
Let me try again...

In your sample code, NUnit considers `Tests` to be a test. It also considers `Test1` to be a test contained in `Tests`.

The implementation of `Test1` is in the `[SetUp]`, `[Test]` and `[TearDown]` methods. If an exception is thrown in one of those methods, NUnit reports it as an error in `Test1`. Since the same `[SetUp]` method is used for all test cases within a fixture, it's possible for the exception to be thrown more than once so, of course, it is reported more than once.

The implementation code for `Tests` is contained in the `[OneTimeSetUp]` and `[OneTimeTearDown]` methods.  If an exception is thrown in either of those methods, NUnit reports it as an error in the test named `Tests`. Both message and stack trace are included. __In addition__, NUnit reports the same error message for each contained test (for `Test1` in this case) as an indicator that you should look at the containing test (fixture `Tests`). Think of this report as a sort of footnote saying "See appendix." That is, the actual error didn't occur in `Test1`, because `Test1` never executed, you need to see the results of `Tests` to understand why.

Unfortunately, you can't see that! The NUnit3 VS Adapter has the job of translating NUnit's view of the tests to Test Explorer's. In Test Explorer's view, only `Test1` is a test. There is no obvious way for the adapter to pass back the results of `Tests` because it is never asked about it.  That's the difficulty in translating between two entirely different views of what a "test" is. 

There doesn't seem to me to be anything more that NUnit itself needs to do. The information is there and is made visible by other runners, like nunit3-console and testcentric-gui. In your case, the actual runner is the dotnet CLI, making use of the adapter. The adapter __could__ do something, but that would probably have to vary depending on the execution environment. When running under the IDE, it could dump more info to the Output window - although that might get very cluttered. When running under vstest or a dotnet command it might write to stdout. In any case, it would have to bypass the normal reporting mechanism to get the information out about a failure that takes place in no recognized "test".

**Comment by provegard** [comment]
Ok, I understand now. Thank you for the detailed explanation! I'll report the problem against the other involved components (Test explorer/Resharper/dotnet tooling).

**Comment by CharliePoole** [comment]
Or the NUnit3 VS adapter, to which we could transfer this issue if you prefer.

**Comment by provegard** [comment]
Aha, I see now that I missed the part about a possible fix/workaround in the adapter. Yes, moving the issue there would be good. Would you like me to do it?
